guac_core/src/storage/mod.rs

```diff
@@ -14,13 +20,23 @@ lazy_static! {
     };
 }
 
+/// Storage contains a futures aware RwLock (QrwLock) which controls access to the inner data
+/// This outer Rwlock should only be mutated very rarely, only to insert and remove counterparties
 pub struct Storage {
     inner: QrwLock<Data>,
 }
 
 #[derive(Default)]
 struct Data {
```

Comment:
I find this and `Storage` insanely vague name-wise. Is this all *data* we will ever intend to use? And the only way we intend to *store* stuff?

Reply:
Also most importantly: Why the hell does this  thing need its own type? And why the hell do you tie two unrelated data structures with one mutex?

Reply:
This is the primary state storage of guac, and it should contain most if not all global state. We need a separate type so we can implement convenience methods on top of it and the 2 separate data structures are needed to effectively create a map from eth address to both the counterparty and channel manager. The 2 maps are tied with one rwlock because they need to stay in sync, and there should be very little reason for getting a write lock on one without getting a write lock on the other.